KAFKA-9216: Enforce that Connect’s internal topics use `compact` cleanup policy

rhauch opened the pull request:
Supplements #8270 

This change adds a check to the KafkaConfigBackingStore, KafkaOffsetBackingStore, and KafkaStatusBackingStore to use the admin client to verify that the internal topics are compacted and do not use the `delete` cleanup policy.

Connect already will create the internal topics with `cleanup.policy=compact` if the topics do not yet exist when the Connect workers are started; the new topics are created always as compacted, overwriting any user-specified `cleanup.policy`. However, if the topics already exist the worker did not previously verify the internal topics were compacted, such as when a user manually creates the internal topics before starting Connect or manually changes the topic settings after the fact.

The current change helps guard against users running Connect with topics that have delete cleanup policy enabled, which will remove all connector configurations, source offsets, and connector & task statuses that are older than the retention time. This means that, for example, the configuation for a long-running connector could be deleted by the broker, and this will cause restart issues upon a subsequent rebalance or restarting of Connect worker(s).

Connect behavior requires that its internal topics are compacted and not deleted after some retention time. Therefore, this additional check is simply enforcing the existing expectations, and therefore does not need a KIP.

Added unit tests for the new logic, and added an integration test that verifies that the worker will fail to start if each of the three internal topics does not use only `cleanup.policy=compact`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

rhauch committed:
KAFKA-9216: Enforce that Connect’s internal topics use `compact` cleanup policy

This change adds a check to the KafkaConfigBackingStore, KafkaOffsetBackingStore, and KafkaStatusBackingStore to use the admin client to verify that the internal topics are compacted and do not use the `delete` cleanup policy.

Connect already will create the internal topics with `cleanup.policy=compact` if the topics do not yet exist when the Connect workers are started; the new topics are created always as compacted, overwriting any user-specified `cleanup.policy`. However, if the topics already exist the worker did not previously verify the internal topics were compacted, such as when a user manually creates the internal topics before starting Connect or manually changes the topic settings after the fact.

The current change helps guard against users running Connect with topics that have delete cleanup policy enabled, which will remove all connector configurations, source offsets, and connector & task statuses that are older than the retention time. This means that, for example, the configuation for a long-running connector could be deleted by the broker, and this will cause restart issues upon a subsequent rebalance or restarting of Connect worker(s).

Connect behavior requires that its internal topics are compacted and not deleted after some retention time. Therefore, this additional check is simply enforcing the existing expectations, and therefore does not need a KIP.

rhauch committed:
KAFKA-9216: Added logic to skip this check if the topic cleanup policy could not be returned

If the Kafka principal that Connect uses for internal topics does not have proper ACLs or the Kafka broker is older than 0.11.0.0, the worker will not be able to validate the cleanup policy for the internal topics. This should not cause the worker to fail.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java:
```diff
@@ -94,6 +95,7 @@
     protected final StatusBackingStore statusBackingStore;
     protected final ConfigBackingStore configBackingStore;
     private final ConnectorClientConfigOverridePolicy connectorClientConfigOverridePolicy;
+    protected final AtomicBoolean running = new AtomicBoolean(false);
```
`volatile boolean` is equivalent to `AtomicBoolean` and if we are not going to use any of the compare-and-set/get-and-set capabilities of the atomic class, maybe you'd want to consider using `volatile` instead to avoid the boilerplate of calling get/set on that boolean variable. But of course, the decision is a matter of style too.

rhauch replied:
ack

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java:
```diff
@@ -492,7 +493,14 @@ public void putSessionKey(SessionKey sessionKey) {
                 try (TopicAdmin admin = new TopicAdmin(adminProps)) {
-                    admin.createTopics(topicDescription);
+                    // Create the topic if it doesn't exist
+                    Set<String> newTopics = admin.createTopics(topicDescription);
```
Vararg gives some unintended consequences in naming. Should we stick to singular given that we expect at most a single topic to be created?

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:
```diff
@@ -375,6 +383,152 @@ public boolean createTopic(NewTopic topic) {
+        String expectedPolicyStr = String.join(",", expectedPolicies);
+        if (cleanupPolicies != null && !cleanupPolicies.equals(expectedPolicies)) {
+            String msg = String.format("Topic '%s' supplied via the '%s' property is required "
+                                       + "to have '%s=%s' to guarantee consistency and durability of "
+                                       + "%s, but found '%s'. "
+                                       + "Correct the topic before restarting Connect.",
```
nit: should we keep 2 tabs alignment for everything?

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:
```diff
@@ -375,6 +383,152 @@ public boolean createTopic(NewTopic topic) {
+     * Get the cleanup policy for a topic.
+     *
+     * @param topic the name of the topic
+     * @return the set of cleanup policies set for the topic; may be empty if the topic exists
```
I wonder if distinguishing like that, with `null` and `empty` pays off. 
Why not return an empty collection in both cases and simplify the checks on the return values of this method?

This method doesn't seem to be the one to use when somebody wants to determine whether a topic exists or not.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:
```diff
@@ -375,6 +383,152 @@ public boolean createTopic(NewTopic topic) {
+     *         thread is interrupted while attempting to perform this operation
+     * @throws ConnectException if a non retriable error occurs
+     */
+    public Map<String, Config> describeTopicConfigs(String...topicNames) {
```
nit: don't we need a space between the varargs type and the variable name? I'm surprised mainly at checkstyle here.

kkonstantine reviewed (commented):
Thanks @rhauch 
Good to add some more checks around internal topics. 
Left a few minor comments, but this looks good overall.

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:
```diff
@@ -375,6 +383,152 @@ public boolean createTopic(NewTopic topic) {
+    public boolean verifyTopicCleanupPolicyOnlyCompact(String topic, String workerTopicConfig,
+            String topicPurpose) {
+        Set<String> cleanupPolicies = topicCleanupPolicy(topic);
+        if (cleanupPolicies == null || cleanupPolicies.isEmpty()) {
```
Is it possible that this will also be true if there isn't a cleanup policy configured on the topic?

rhauch replied:
Is it even possible to have a blank `cleanup.policy`? The topic will inherit the broker's `log.cleanup.policy`, which has a default and which is not allowed to be blank. And it looks like `cleanup.policy` is required to be either `compact` or `delete` or both.

C0urante replied:
No kidding... I assumed it was possible to create topics without cleanup policies but it looks like you're right. My bad!

rhauch replied:
After some testing, I've discovered that even if the topic-specific settings don't explicitly specify the `cleanup.policy`, the config description for the topic returned by the admin client will still include `cleanup.policy=<broker's log.cleanup.policy>`. That means if we have permissions and the topic exists, we will always get the `cleanup.policy` for the topic.

And, I've cleaned up the semantics of the `topicCleanupPolicy(...)` method to never return a null set. This means that if the set is empty, either the topic doesn't exist or the admin client could not return the topic settings (e.g., old broker, insufficient ACLs, etc.). Either way, this logic gets simplified a bit.

I hope that helps address your concern/question.

C0urante replied:
I think my concern was invalid to begin with but your refactoring is certainly an improvement. LGTM 👍

rhauch committed:
KAFKA-9216: Used volatile instead of AtomicBoolean

rhauch committed:
KAFKA-9216: Incorporated review feedback by improving semantics

rhauch committed:
KAFKA-9216: Corrected test case

rhauch commented:
@kkonstantine and @C0urante: thanks for the review. I think I've incorporated all of your feedback and addressed all of your questions. @C0urante, I've even tried to improve the failure message to say what needs to be done if the topic has an unacceptable `cleanup.policy`.

I'd appreciate another pass. Thanks!

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:
```diff
@@ -375,6 +383,162 @@ public boolean createTopic(NewTopic topic) {
+            String topicPurpose) {
+        Set<String> cleanupPolicies = topicCleanupPolicy(topic);
+        if (cleanupPolicies.isEmpty()) {
+            log.debug("Unable to use admin client to verify the cleanup policy of '{}' "
```
Should we consider `info`? It's a one time message right?

rhauch replied:
I could see changing this to `info`, because this is important. But the others log messages really are just tracking that we're using the admin client and what we're finding, so I think `debug` is probably the best there. If the cleanup policy is wrong, then if we're logging that we're also going to fail the worker; if the cleanup policy is acceptable, I don't think it's worth logging it at `info`.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:
```diff
@@ -375,6 +383,162 @@ public boolean createTopic(NewTopic topic) {
+        ConfigEntry entry = topicConfig.get(CLEANUP_POLICY_CONFIG);
+        if (entry != null && entry.value() != null) {
+            String policyStr = entry.value();
+            log.debug("Found cleanup.policy={} for topic '{}'", policyStr, topic);
```
same question around log level as above

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:
```diff
@@ -375,6 +383,162 @@ public boolean createTopic(NewTopic topic) {
+        }
+        // This is unexpected, as the topic config should include the cleanup.policy even if
+        // the topic settings don't override the broker's log.cleanup.policy. But just to be safe.
+        log.debug("Found no cleanup.policy for topic '{}'", topic);
```
same question around log level as above

kkonstantine reviewed (approved):
Thanks @rhauch 
As a last thought, I wonder if we want to elevate a few of the log messages at the `info` level to proactively give a hint to the users. Or you may leave them at `debug`. I'm fine either way, just wanted to call out this last aspect. 

LGTM!

rhauch committed:
KAFKA-9216: Changed one log message to info
